src/TorchSharp/Distributions/Constraints.cs

```diff
@@ -0,0 +1,637 @@
+                    public virtual bool is_discrete { get; protected set; }
+
+                    /// <summary>
+                    /// Number of rightmost dimensions that together define an event.
+                    /// The check() method will remove this many dimensions when computing validity.
+                    /// </summary>
+                    public virtual int event_dim { get; protected set; }
+
+                    public override string ToString()
+                    {
+                        return this.GetType().Name + "()";
+                    }
+                }
+
+                /// <summary>
+                /// Placeholder for variables whose support depends on other variables.
+                /// These variables obey no simple coordinate-wise constraints.
+                /// </summary>
+                public class _Dependent : Constraint
+                {
+                    public _Dependent(bool is_discrete = false, int event_dim = 0) : base(is_discrete, event_dim) { }
+
+                    public override Tensor check(Tensor value)
+                    {
+                        throw new ArgumentException("Cannot determine validity of dependent constraint");
+                    }
+                }
+
+                /// <summary>
+                /// Wraps a constraint by aggregating over ``reinterpreted_batch_ndims``-many
+                /// dims in :meth:`check`, so that an event is valid only if all its
+                /// independent entries are valid.
+                /// </summary>
+                public class _IndependentConstraint : Constraint
+                {
+                    public _IndependentConstraint(Constraint base_constraint, int reinterpreted_batch_ndims)
+                    {
+
+                    }
+
+                    public override Tensor check(Tensor value)
+                    {
+                        throw new NotImplementedException();
+                    }
+
+                    public override string ToString()
+                    {
+                        return this.GetType().Name + "()";
+                    }
+
+                    public Constraint base_constraint { get; private set; }
+
+                    /// <summary>
+                    /// Whether constrained space is discrete.
+                    /// </summary>
+                    public override bool is_discrete { get => base_constraint.is_discrete; }
+
+                    /// <summary>
+                    /// Number of rightmost dimensions that together define an event.
+                    /// The check() method will remove this many dimensions when computing validity.
+                    /// </summary>
+                    public override int event_dim { get => base_constraint.event_dim; }
+
+                }
+
+                /// <summary>
+                /// Constrain to the two values {0, 1}.
+                /// </summary>
+                public class _Boolean : Constraint
```

Comment:
>_Boolean [](http://example.com/codeflow?start=29&length=8)

Is it how it is defined with pytorch (started with underscore)?

Reply:
Yes. I held my .NET naming conventions nose and called everything the same as Pytorch does.